### ISSUE TITLE
Setting additional HTTP properties on responses (ie. cookies & headers)

### ISSUE THREAD

**Issue by naamancurtis** [async-graphql/async-graphql#370]
## Description of the feature

Is there any interest (or current ability to) support setting cookies/headers on the response for a query or mutation?

The example I'm thinking of would be access/refresh tokens that would be stored in `HttpOnly` cookies as a result of hitting a `refresh` mutation. I must admit I'm still relatively new to GraphQL, so the answer might just be to remove auth related functionality from GraphQL and do it via some other API technology, but I figured I'd ask.

Looking at the code, I imagine `async_graphql::Response` would have to be extended to support this?

**Comment by sunli829** [comment]
> ## Description of the feature
> Is there any interest (or current ability to) support setting cookies/headers on the response for a query or mutation?
> 
> The example I'm thinking of would be access/refresh tokens that would be stored in `HttpOnly` cookies as a result of hitting a `refresh` mutation. I must admit I'm still relatively new to GraphQL, so the answer might just be to remove auth related functionality from GraphQL and do it via some other API technology, but I figured I'd ask.
> 
> Looking at the code, I imagine `async_graphql::Response` would have to be extended to support this?

I recommend sharing `Arc<Mutex<Option<String>>>` through `Context`.

Following is the example:

```rust
let cookie: Arc<Mutex<Option<String>>> = Default::default();

schema.execute(request.data(cookie)).await;

#[Object]
implement Mutation {
    async refresh_token(&self, ctx: &Context<'_>, ...) -> bool {
        *ctx::data_unchecked::<Arc<Mutex<Option<String>>>>().lock().await = Some(new_refresh_token);
        true
    }
}
```

**Comment by scrogson** [comment]
I believe the question is regarding setting cookies on the HTTP response.

**Comment by naamancurtis** [comment]
Yep, setting cookies/headers on the HTTP response ideally covering both use cases 
1. Setting brand new cookies/headers that don't currently exist on the request.
2. Updating existing cookies/headers with new values.

I guess the challenge here is each integration (ie. actix-web, warp etc) has its own notion of cookies & headers _(I think?)_ and the conversion from 'async_graphql::Response' to 'framework_library::HttpResponse' happens within the integration library if I understand the flow correctly?

**Comment by sunli829** [comment]
This example is based on `actix-web`, and explains more fully how to get the response HTTP header from the `GraphQL` query and return it to the client.

```rust

struct;

#[Object]
impl Mutation {
    async refresh_token(&self, ctx: &Context<'_>, ...) -> bool {
        *ctx::data_unchecked::<Arc<Mutex<Option<String>>>>().lock().await = Some(new_refresh_token);
        true
    }
}

async fn index(schema: web::Data<MySchema>, req: Request) -> Response {
    let cookie: Arc<Mutex<Option<String>>> = Default::default();
    let resp: Response = schema.execute(req.into_inner().data(cookie)).await.into();
    if let Some(cookie_value) = cookie.lock().await {
        resp.with_header("Cookie", cookie_value);
    }
    resp
}
```

**Comment by naamancurtis** [comment]
I'm possibly just doing something wrong, but the types got quite rough when going with this approach, the 4 response types (`actix_web::HttpResponse`, `actix_web::CustomResponder`, `async_graphql::Response` and `async_graphql_actix_web::Response`) and the `actix_web::Responder` trait made things a bit tricky. Especially when you want to set more than 1 header like this, although I did manage to get it to compile.

I think for your snippet above you'd need to type the response type of `index` as `impl Responder`? as `Responder.with_header()` returns an `actix_web::CustomerResponder` which isn't exposed publically?

I did end up getting something to work which made slightly more sense to me, but I had to dip into the async_graphql_actix_web code to do so. I don't know if this can be tidied up/improved somehow?

```rust
pub async fn graphql_index(
    schema: web::Data<BazaarSchema>,
    http_request: HttpRequest,
    graphql_request: Request,
) -> Result<HttpResponse> {
    let access_cookie = http_request
        .cookie(TokenType::Access.as_str())
        .map(|c| c.value().to_string());
    let refresh_cookie = http_request
        .cookie(TokenType::Refresh(0).as_str())
        .map(|c| c.value().to_string());

    let cookies = Arc::new(Cookies::new(access_cookie, refresh_cookie)?);

    let mut request = graphql_request.into_inner();
    request = request.data(Arc::clone(&cookies));

    let res: Response = schema.execute(request).await.into();
    let mut response = HttpResponse::build(StatusCode::OK);
    response.content_type("application/json");
    if res.0.is_ok() {
        if let Some(cache_control) = res.0.cache_control().value() {
            response.header("cache-control", cache_control);
        }
    }

    set_cookie_on_response(
        &mut response,
        cookies.get_access_cookie().ok().flatten(),
        TokenType::Access,
    );
    set_cookie_on_response(
        &mut response,
        cookies.get_refresh_cookie().ok().flatten(),
        TokenType::Refresh(0),
    );
    response.json(&res.0).await
}
```

**Comment by sunli829** [comment]
This really looks bad, so I decided to add the `headers` attribute to `Response`. 😀

**Comment by sunli829** [comment]
Released in `v2.4.6`.

**Comment by naamancurtis** [comment]
Thanks for turning this around so quickly, I think there are some slight limitations with a `HashMap` approach, whereby there is no way for you to have multiple values for the same key, which is supported by the Headers spec. 

I've made a small PR just to tweak the implementation slightly to use the `http::HeaderMap` implementation, hopefully this works for you?

**Comment by naamancurtis** [comment]
Resolved by #370 & #383
